compiler/src/iree/compiler/Dialect/Flow/Transforms/SinkReshapes.cpp

```diff
@@ -64,35 +78,79 @@ static bool shouldSinkExpandShapeOp(OpOperand *opOperand) {
-               // The check for the producer having a single use is not fully
-               // worked out. Ideally we can fuse with a producer irrespective
-               // of number of uses, but is a good thumb rule in practice.
-               llvm::hasSingleElement(currProducer->getUses());
-      })) {
+  // First check that the expand_shape producer and consumer can be fused.
+  Operation *reshapeProducer = reshapeOp.getSrc().getDefiningOp();
+  if (!reshapeProducer) {
     return false;
   }
-
-  // Do not sink if consumer is a contraction/matmul like op.
-  if (auto linalgConsumerOp = dyn_cast<linalg::LinalgOp>(consumer)) {
-    if (linalg::isaContractionOpInterface(linalgConsumerOp))
-      return false;
+  if (!isFusableUsingTileAndFuse(reshapeOp.getSrc().getDefiningOp(),
+                                 consumer)) {
+    return false;
   }
 
-  return isFusableUsingTileAndFuse(reshapeOp.getSrc().getDefiningOp(),
-                                   consumer);
+  // If the op is already fusable with producer using tile and fuse,
+  // do nothing.
+  for (OpOperand &opOperand : consumer->getOpOperands()) {
+    Operation *currProducer = opOperand.get().getDefiningOp();
+    if (!currProducer) {
+      continue;
+    }
+
+    // The check for the producer having a single use is not fully
+    // worked out. Ideally we can fuse with a producer irrespective
+    // of number of uses, but is a good thumb rule in practice.
+    if (!llvm::hasSingleElement(currProducer->getUses())) {
+      continue;
+    }
+
+    // Check if a producer can already be tiled and fused with the consumer.
+    if (!isFusableUsingTileAndFuse(currProducer, consumer)) {
+      continue;
+    }
+
+    // There is already a tile-and-fusable producer to fuse with. Still prefer
+    // fusing with the producer whose parallel iteration space rank matches
+    // the consumer parallel iteration space rank to avoid loss of parallelism.
+    if (isa<linalg::LinalgOp>(reshapeProducer) &&
+        isa<linalg::LinalgOp>(currProducer)) {
+      unsigned currConsumerNumParallelLoops =
+          consumerGenericOp.getNumParallelLoops();
+      unsigned currProducerNumParallelLoops =
+          cast<linalg::LinalgOp>(currProducer).getNumParallelLoops();
+      if (currProducerNumParallelLoops == currConsumerNumParallelLoops) {
+        // If the producer has same number of parallel loops as consumer,
+        // then this is the operand to fuse along. So do nothing.
+        return false;
+      }
+      // If the producer has less number of parallel loops as the consumer,
+      // ignore this operand.
+      if (currProducerNumParallelLoops < currConsumerNumParallelLoops) {
+        continue;
+      }
+      unsigned reshapeProducerNumParallelLoops =
+          cast<linalg::LinalgOp>(reshapeProducer).getNumParallelLoops();
+      if (currProducerNumParallelLoops < reshapeProducerNumParallelLoops) {
+        return false;
+      }
```

Comment:
When we compare the reshape producer's parallel loops, shouldn't we account for the expand_shape? The consumer has only parallel loops, so we should be able to infer the number of parallel loops in the collapsed domain. We could compare `currProducerNumParallelLoops - consumerNumParallelLoops` with `reshapeProducerNumParallelLoops - collapsedConsumerNumParallelLoops` to see which one has a larger mismatch between number of parallelizable dimensions.

Reply:
I dont think so (but I might be wrong)... If the reshape is pushed down then the `reshapeProducerNumParallelLoops` still remains the same (its the producer of the reshape that is untouched). So the comparison is between that and what the `currProducerNumParallelLoops` are?